### PR TITLE
Occupation Classifier pipeline components

### PR DESCRIPTION
**Pull request by tweddielin**
I see the whole pipeline break down into `Embedding`, `Classifier`, `Logger` and `Test`. This PR is addressing the `Embedding` and `Classifier`.

- `CombinedClassifier`: a class to combine embedding and classifier using scikit-learn pipeline object 
- `OccupationClassifierTrainer`: a trainer to train and cross-validate a series of classifier given the conifg file. Now the cross-validation result is the property, but we will want to convert it into a json format as well as the testing suite. 
- `EmbeddingTransformer`: scikit-learn interface for embedding model
- `create_training_set`: a helper function to create training data
- `TrainingMatrix`: a class holds training matrix, label matrix, embedding model and target variable to feed in a trainer
- `TargetVariable`: a class has abstract property methods of name and transformer, and take self-defined filters as input.
- `SOCMajorGroup`: is inherited from `TargetVariable` and has default properties of SocEncoder and two soc code filters(`unknow_soc_filter` and `empty_soc_filter`).

### CODE REVIEW

File: skills_ml/algorithms/occupation_classifiers/classifiers.py [hunk 1]
```diff
@@ -8,29 +14,56 @@
...
 
+class CombinedClassifier(object):
+    def __init__(self, embedding, classifier, **kwargs):
+        self.embedding = embedding
+        self.classifier = classifier
+        self.combined = Pipeline([
+            ('tokens_to_vector', EmbeddingTransformer(self.embedding)),
+            ('classify', self.classifier)
+        ])
+        self.soc_encoder = SocEncoder(list(majorgroupname.keys()))
```
Comment (thcrock) [inline skills_ml/algorithms/occupation_classifiers/classifiers.py]: Is this tying the classifier too closely to the major group? In the training code there are separate code paths for major group and full soc but here it's hardcoded.

File: skills_ml/algorithms/occupation_classifiers/train.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        document_schema_fields (list): fields to be included in the training data
+
+    Returns:
+        (dict) a dictionary of training data(X), label(y), the embedding_model, target_variable and soc_encoder
+    """
+    X = []
+    y = []
+
+    if target_variable == "major_group":
+        se = SocEncoder(list(majorgroupname.keys()))
```
Comment (thcrock) [inline skills_ml/algorithms/occupation_classifiers/train.py]: Not totally necessary to do any of this, but my thoughts on this section:

I think it would be cleaner to define these lists higher up, since it doesn't seem like it's just the trainer that uses them. At module level (like in occupation_classifiers.__init__.py)

Even better: Since the list and label transformer are closely coupled, put them in the same thing!

```
class TargetVariable(object):
  abstract name

class SOCMajorGroup(TargetVariable):
    name = 'major_group'

    def __init__(self):
        self.choices = list(majorgroupname.keys())
        self.encoder = SocEncoder(self.choices)
        self.transformer = lambda job_posting: self.encoder.transform([job_posting['onet_soc_code'][:2]])

```

With a little more thought you could even fit the filtering (if soc code doesn't start with 99, etc) into this framework.

File: skills_ml/algorithms/occupation_classifiers/classifiers.py [hunk 3]
```diff
@@ -1,5 +1,11 @@
 from skills_ml.algorithms.embedding.base import ModelStorage
-from skills_ml.algorithms.embedding.models import Doc2VecModel
+from skills_ml.algorithms.embedding.models import Doc2VecModel, EmbeddingTransformer
+from skills_ml.algorithms.occupation_classifiers.train import SocEncoder
+from skills_ml.ontologies.onet import majorgroupname
+
+from sklearn.base import BaseEstimator, ClassifierMixin
+from sklearn.utils.validation import check_X_y, check_array, check_is_fitted
+from sklearn.pipeline import make_pipeline, Pipeline
```
Comment (thcrock) [inline skills_ml/algorithms/occupation_classifiers/classifiers.py]: Could clean up the imports. I was curious about some of these imports and Ctrl+F'd them to no avail.

File: skills_ml/algorithms/occupation_classifiers/train.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                module = importlib.import_module(module_name)
+                cls = getattr(module, class_name)
+                kf = StratifiedKFold(n_splits=self.k_folds, random_state=self.random_state_for_split)
+
+                cls_cv = GridSearchCV(cls(), parameter_config, cv=kf, scoring=score, n_jobs=self.n_jobs)
+                cls_cv.fit(X, y)
+                self.cls_cv_result[score] = {class_name: cls_cv.cv_results_}
+                self.best_classifiers[score] = {class_name: cls_cv}
+
+    def save_result(self):
```
Comment (thcrock) [inline skills_ml/algorithms/occupation_classifiers/train.py]: What is the plan for saving the classifiers we want to use? We have a storage object all ready to go and what looks like some classifiers to save.

File: skills_ml/algorithms/occupation_classifiers/train.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                }
+
+    def train(self):
+        """Fit a model to a training set. Works on any modeling class that
+        is vailable in this package's environment and implements .fit
+        """
+        X = self.matrix['X']
+        y = self.matrix['y']
+        for score in self.scores:
+            for class_path, parameter_config in self.grid_config.items():
```
Comment (thcrock) [inline skills_ml/algorithms/occupation_classifiers/train.py]: This loop is the part that takes a while, so we probably want to log a bunch of things

File: skills_ml/algorithms/occupation_classifiers/train.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            label = label_transformer(job['onet_soc_code'])
+            text = ' '.join([job[d] for d in document_schema_fields])
+            tokens = NLPTransforms().word_tokenize(text)
+            if embedding_model:
+                X.append(embedding_model.infer_vector(tokens))
+            else:
+                X.append(tokens)
+            y.append(label)
+
+    return {'X': np.array(X),
```
Comment (thcrock) [inline skills_ml/algorithms/occupation_classifiers/train.py]: Would be nice to log the training set stats (especially the size) before returning.

File: skills_ml/algorithms/occupation_classifiers/train.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+    if target_variable == "major_group":
+        se = SocEncoder(list(majorgroupname.keys()))
+        label_transformer = lambda soc_code: se.transform([soc_code[:2]])
+    elif target_variable == "full_soc":
+        se = SocEncoder(get_all_soc())
+        label_transformer = lambda soc_code: se.transform([soc_code])
+
+    for job in job_postings_generator:
+        if job['onet_soc_code'][:2] != '99' and job['onet_soc_code'] != '' :
```
Comment (thcrock) [inline skills_ml/algorithms/occupation_classifiers/train.py]: Maybe keep track of how many got skipped, and log it at the end so we can see what percentage of the corpus got dropped on the floor before training.

File: skills_ml/algorithms/occupation_classifiers/train.py (hunk 4 above)
Comment (tweddielin) [reply skills_ml/algorithms/occupation_classifiers/train.py]: I don't think we're gonna save the model/classifier itself but the cross-validation results and model hyper-parameters which is going to be in another PR. It's gonna be like what we've been talking about that in the json format.

Comment (thcrock) [reply skills_ml/algorithms/occupation_classifiers/train.py]: I do remember talk about this, but maybe I misunderstood.

Ultimately we'll want to use the winner/s for prediction on a large dataset. I don't want to make the code that does this crawl through job postings again.

Is it that we will save the training matrix, so that usage just involves loading that matrix, the hyperparameters, and calling .fit?

Comment (tweddielin) [reply skills_ml/algorithms/occupation_classifiers/train.py]: Ah right. sorry I made it confusing. It's not that we don't want to save the winner classifier but maybe not in this stage. The winner classifier in this stage is the winner in training/validation. It might be over-fitting to become the winner, so there will be a testing stage to see how it performs on our test dataset(e.g. our uniform 24k sample). I'm just not sure if it's the best thing to do to save the classifier right after the training stage.  The other question would be if we want to save classifier do we want to save all or just the winner in this stage.
